File: typescript/src/sse.ts
/**
 * Server-Sent Events (SSE) stream parser.
 */

export interface SSEEvent {
  /** Event type (defaults to 'message' per SSE spec). */
  event: string;
  /** Event data (may contain newlines from multiple data: lines). */
  data: string;
  /** Optional event ID. */
  id?: string;
}

/**
 * Parse a ReadableStream of bytes into SSE events.
 *
 * Follows the standard SSE protocol:
 * - Lines starting with `event:` set the event type
 * - Lines starting with `data:` append to the data buffer
 * - Lines starting with `id:` set the event ID
 * - Empty lines dispatch the buffered event
 * - Lines starting with `:` are comments (ignored)
 *
 * @param stream - A ReadableStream of Uint8Array chunks (e.g. from fetch response.body).
 * @yields Parsed SSE events.
 */
export async function* parseSSEStream(
  stream: ReadableStream<Uint8Array>
): AsyncGenerator<SSEEvent> {
  const decoder = new TextDecoder();
  const reader = stream.getReader();
  let buffer = '';
  let eventType = '';
  let dataLines: string[] = [];
  let eventId: string | undefined;

  try {
    while (true) {
      const { done, value } = await reader.read();
      if (done) break;

      buffer += decoder.decode(value, { stream: true });

      // Process complete lines
      let newlineIdx: number;
      while ((newlineIdx = buffer.indexOf('\n')) !== -1) {
        const line = buffer.slice(0, newlineIdx).replace(/\r$/, '');
        buffer = buffer.slice(newlineIdx + 1);

        if (line === '') {
          // Empty line = dispatch event if we have data
          if (dataLines.length > 0) {
            yield {
              event: eventType || 'message',
              data: dataLines.join('\n'),
              ...(eventId !== undefined && { id: eventId }),
            };
          }
          // Reset for next event
          eventType = '';
          dataLines = [];
          eventId = undefined;
        } else if (line.startsWith(':')) {
          // Comment — ignore
        } else if (line.startsWith('event:')) {
          const val = line.slice(6);
          eventType = val.startsWith(' ') ? val.slice(1) : val;
        } else if (line.startsWith('data:')) {
          const val = line.slice(5);
          dataLines.push(val.startsWith(' ') ? val.slice(1) : val);
        } else if (line.startsWith('id:')) {
          const val = line.slice(3);
          eventId = val.startsWith(' ') ? val.slice(1) : val;
        }
      }
    }

    // Process any remaining partial line in the buffer
    if (buffer.length > 0) {
      const line = buffer.replace(/\r$/, '');
      if (line.startsWith('event:')) {
        const val = line.slice(6);
        eventType = val.startsWith(' ') ? val.slice(1) : val;
      } else if (line.startsWith('data:')) {
        const val = line.slice(5);
        dataLines.push(val.startsWith(' ') ? val.slice(1) : val);
      } else if (line.startsWith('id:')) {
        const val = line.slice(3);
        eventId = val.startsWith(' ') ? val.slice(1) : val;
      }
    }

    // Flush any remaining data after stream ends
    if (dataLines.length > 0) {
      yield {
        event: eventType || 'message',
        data: dataLines.join('\n'),
        ...(eventId !== undefined && { id: eventId }),
      };
    }
  } finally {
    reader.releaseLock();
  }
}

File: python/src/mentiora/sse.py
"""Server-Sent Events (SSE) parser utilities."""

from collections.abc import AsyncIterator, Iterator
from dataclasses import dataclass, field


@dataclass
class SSEEvent:
    """A parsed Server-Sent Event."""

    event: str = ''
    data: str = ''
    id: str | None = None


@dataclass
class _SSEBuffer:
    """Accumulator for SSE fields across lines."""

    event: str = ''
    data_lines: list[str] = field(default_factory=list)
    id: str | None = None

    def has_data(self) -> bool:
        return bool(self.data_lines) or bool(self.event)

    def to_event(self) -> SSEEvent:
        return SSEEvent(
            event=self.event or 'message',
            data='\n'.join(self.data_lines),
            id=self.id,
        )

    def reset(self) -> None:
        self.event = ''
        self.data_lines = []
        self.id = None


def _process_line(line: str, buf: _SSEBuffer) -> SSEEvent | None:
    """Process a single SSE line. Returns an event on blank line (dispatch), else None."""
    # Blank line = dispatch event
    if not line:
        if buf.has_data():
            event = buf.to_event()
            buf.reset()
            return event
        return None

    # Comment lines start with ':'
    if line.startswith(':'):
        return None

    # Parse "field: value" or "field:value" or "field"
    if ':' in line:
        field_name, _, value = line.partition(':')
        # Strip single leading space from value per SSE spec
        if value.startswith(' '):
            value = value[1:]
    else:
        field_name = line
        value = ''

    if field_name == 'event':
        buf.event = value
    elif field_name == 'data':
        buf.data_lines.append(value)
    elif field_name == 'id':
        buf.id = value

    return None


def parse_sse_lines(lines: Iterator[str]) -> Iterator[SSEEvent]:
    """Parse SSE events from a synchronous line iterator.

    Args:
        lines: Iterator of text lines (without trailing newlines).

    Yields:
        Parsed SSEEvent objects.
    """
    buf = _SSEBuffer()
    for line in lines:
        event = _process_line(line, buf)
        if event is not None:
            yield event
    # Flush any remaining buffered event at end of stream
    if buf.has_data():
        yield buf.to_event()


async def parse_sse_lines_async(lines: AsyncIterator[str]) -> AsyncIterator[SSEEvent]:
    """Parse SSE events from an asynchronous line iterator.

    Args:
        lines: Async iterator of text lines (without trailing newlines).

    Yields:
        Parsed SSEEvent objects.
    """
    buf = _SSEBuffer()
    async for line in lines:
        event = _process_line(line, buf)
        if event is not None:
            yield event
    # Flush any remaining buffered event at end of stream
    if buf.has_data():
        yield buf.to_event()
